Friendlier use of Redis

kevindew opened the pull request:
Based off this card: https://trello.com/c/o0i6SGFf/794-experiments-framework-fix-medium and a chat with @steventux 

This PR covers 3 aspects and can be broken down if any of these require extra discussion.
### Set expiry dates on all data going into redis

Each time we set data in redis we set an expiry on it to allow redis to auto clean-up that data.
By default this is set to 60 seconds for candidate data and 24 hours for a result. This can be configured with arguments to the `experiment_control` method:

``` ruby
experiment_control(..., results_expiry: 60 * 5, candidate_expiry: 60)
```

We also only retry 5 times in the control worker to see if candidate data is ready otherwise we abort.
### Update documentation

I've added some more information on what this thing is, what it does and how to set it up - with a link to a working example in publishing-api
### Less duplicate data in redis

In the cases where we're storing mismatches and errors I've switched it to store a sha2 hash of the data as part of the key and not add new entries into redis when we already have one item matching that hash. This is to reduce the amount of data that is redundantly stored in redis.

@elliotcm as original author I thought best to tag you in so you get a notification.

kevindew committed:
Add expiry to all data added to redis

To reduce the amount of data stored in redis and for how long it is
stored this adds an expiry to all data entered into redis by this gem.

By default a candidate results will be stored for 60 seconds and the
results are stored for 24 hours.

This can be customised with the hash argument to `experiment_control`

eg `experiment_control(..., results_expiry: 60 * 5, candidate_expiry: 60)`

kevindew committed:
Support a run_output of nil

Previously if the output of a run was nil there was an assumption that
the run output hadn't been retrieved. A presence of a duration is
sufficient as this should be non-nil.

kevindew committed:
Update documentation

Extra information to make it easier for someone to get moving with this
gem.

kevindew committed:
Add expiry documentation

kevindew committed:
Hash results so they are not stored in redis twice

Before data of a mismatch is entered into redis a SHA2 hash is
generated for the data and redis is checked to see if that data already
exists. If it exists the new data is not added and the expiry time is
extended. If it does not exist it is added to redis.

kevindew committed:
Make block formatting consistent with styleguides

kevindew committed:
Only store one version of an error

This hashes errors that are stored to try reduce the risk of storing
multiple instances of an error.
A method `get_experiment_exceptions` has been added to `AsyncExperiments` for
retrieving the list of exceptions for a experiment.

kevindew committed:
Bump version

carlosmartinez commented on README.md:
```diff
@@ -1,10 +1,15 @@
-output for later review.
+This tool is used to understand the implications of replacing a section of code
+(known as control) with a different piece of code (the candidate) in terms of
+interchangeable outputs and affects on performance. It reports differences in
```
`s/affects/effects`

carlosmartinez commented on lib/async_experiments.rb:
```diff
@@ -12,20 +14,18 @@ def self.statsd=(statsd)
-        operator == "-"
-      }
+    mismatched_responses.map do |parsed|
+      missing, other = parsed.partition { |(operator)| operator == "-" }
```
are the brackets needed around `operator`?

kevindew replied:
They're used to access the first item in the array, it might be that there is a more idiomatic way to represent this I'm not sure :-/

carlosmartinez commented on lib/async_experiments.rb:
```diff
@@ -37,7 +37,17 @@ def self.get_experiment_data(experiment_name)
+      )
+      retrieve = -> (key) { redis.get(key) }
+      enumerator.map(&retrieve).compact
+    end
```
might be nice to extract some of this logic that's shared with `get_experiment_data`

carlosmartinez commented on lib/async_experiments/experiment_result_control_worker.rb:
```diff
@@ -0,0 +1,29 @@
+
+    LOCK_TIMEOUT = 60
+
+    def perform(name, id, run_output, duration, expiry, allowed_attempts = 5, attempt = 1)
```
with this many params, is there a case for using named arguments?

kevindew replied:
Good point

kevindew replied:
Oh I remember why I didn't now. You can't pass named arguments into a sidekiq worker, so instead we'd have to pass in a hash and then use `fetch` on it. Similarly it'd be worth keeping the `initialize` interface consistent with `AsyncExperiments::ExperimentResultCandidateWorker`. That still has 5 init arguments though. Feels like it'd add quite a bit of clutter though - so I think it might be better as an optimisation if the list of arguments grows further.

carlosmartinez commented on spec/async_experiments_spec.rb:
```diff
@@ -3,41 +3,48 @@
       expect(results).to eq([
         missing: [
-          {"same_key" => 1, "different_key" => 2},
+          { "same_key" => 1, "different_key" => 2 },
```
the formatting gods are smiling on you

kevindew replied:
![Testify](http://i.giphy.com/3o6MbisQwZAh4Lqesw.gif)

carlosmartinez commented:
![approve15](https://cloud.githubusercontent.com/assets/168365/18382925/08a2e554-767b-11e6-9bb5-4270a22f7244.gif)

kevindew committed:
Move the redis scan and retrieve functionality

Moves this functionality to be handled in a single method rather than
having both `get_experiment_data` and `get_experiment_exceptions`
implement it individually.

kevindew commented:
@carlosmartinez all comments acked. Thanks for the review :)
